Adding login and logout as starchart user from initial seed

sfrunza13 opened the pull request:
Closes #22
Adding login page with login button that signs you (creates a session as per the session server that came with blues stack) in as the user that we seed username:"starchartdev".

humphd commented on app/routes/index.tsx:
```diff
@@ -1,8 +1,26 @@
+export const loader = async ({ request }: LoaderArgs) => {
+  const username = await requireUsername(request);
+
+  return json({ username: username });
```
nit: when the key/value are the same, include it only once:

```ts
return json({ username });
```

humphd commented on app/routes/index.tsx:
```diff
@@ -1,8 +1,26 @@
     <main>
       <h1>Starchart</h1>
       <p>Start making your own custom domains and certificates today!</p>
+      <p>{data.username}</p>
```
Let's use [Chakra UI components](https://chakra-ui.com/docs/components) and [Remix's `<Form>` component](https://remix.run/docs/en/v1/components/form).  Here's a suggestion:

```tsx
<Text>Welcome {data.username}</Text>
<Form action="/logout" method="post">
  <Button type="submit" colorScheme="red">Logout</button>
</Form>
```

humphd commented on app/routes/login.tsx:
```diff
@@ -0,0 +1,21 @@
+export default function Login() {
+  return (
+    <Form method="post">
+      <h1>Login</h1>
```
```tsx
<Button type="submit" colorScheme="red">Login</button>
```

humphd reviewed (changes requested):
This is looking good.  A few things to use newer code, and also you'll need to rebase to get the changes to `index.tsx`

sfrunza13 commented:
> This is looking good. A few things to use newer code, and also you'll need to rebase to get the changes to `index.tsx`

How's it now?

humphd commented on app/routes/index.tsx:
```diff
@@ -1,6 +1,17 @@
-import { Heading, Text } from "@chakra-ui/react";
+import { Heading, Text, Button } from "@chakra-ui/react";
+import { json, LoaderArgs } from "@remix-run/node";
```
`LoaderArgs` is being imported as a type, so you separate it from your other imports and use `import type`.  You do it like this:

```ts
import { json } from "@remix-run/node";

import type { LoaderArgs } from "@remix-run/node";
```

humphd commented on app/routes/index.tsx:
```diff
@@ -9,6 +20,12 @@ export default function Index() {
         Start making your own custom domains and certificates today!
       </Text>
+      <Text>Welcome {data.username}</Text>
+      <form action="/logout" method="post">
```
`<form>` is HTML, you want `<Form>` from Remix.

humphd commented on app/routes/login.tsx:
```diff
@@ -0,0 +1,27 @@
+import { Heading, Text, Button } from "@chakra-ui/react";
+import { Form } from "@remix-run/react";
+import { ActionArgs } from "@remix-run/node";
```
Same with this: move to a new line, `import type ...`

sfrunza13 commented:
Hopefully that does it

humphd commented on app/routes/index.tsx:
```diff
@@ -1,6 +1,18 @@
-import { Heading, Text } from "@chakra-ui/react";
+import { Heading, Text, Button } from "@chakra-ui/react";
+import { json } from "@remix-run/node";
+import type { LoaderArgs } from "@remix-run/node";
```
Move your type imports below your regular imports and separate with a blank line:

```ts
import {...} from "..."
import {...} from "..."

import type {...} from "..."
import type {...} from "..."
```

humphd commented on app/routes/login.tsx:
```diff
@@ -0,0 +1,25 @@
+import { Heading, Text, Button } from "@chakra-ui/react";
+import { Form } from "@remix-run/react";
+import type { ActionArgs } from "@remix-run/node";
```
Same thing here.

sfrunza13 commented:
I put the types under other imports and squashed, unfortunately realized I had not ordered things right in the logout and added another commit to fix the order of the imports there too

humphd commented:
You have conflicts in `app/routes/index.tsx`, you need to rebase and fix.

sfrunza13 commented:
> You have conflicts in `app/routes/index.tsx`, you need to rebase and fix.

is it correct to do as I am doing and pull upstream into origin main, then checkout my feature branch and rebase main and force push?

humphd commented:
```
git checkout main
git pull upstream main
git checkout your-branch
git rebase main
git push origin your-branch -f
```

sfrunza13 commented:
Cool, that's what I did :D

sfrunza13 commented:
and after all that's done and changes are approved is it ok to just hit squash and merge here in the PR?

humphd commented:
Not sure why Prettier is failing here, but your code changes look good.

sfrunza13 commented:
> Not sure why Prettier is failing here, but your code changes look good.

this has made me realize there was a format script to rewrite with prettier before my last rebase and now there isn't

Eakam1007 commented:
Looks good. Before merging, you can use `npx prettier --write .` to rewrite with prettier

Eakam1007 reviewed (approved):
Looks good! 
Is there an issue to add end-to-end tests for this once #85 is merged? This page will likely change in the future but I think we should start adding tests early

sfrunza13 commented:
> Looks good! Is there an issue to add end-to-end tests for these pages once #85 is merged? This page will likely change in the future but I think we should start adding tests early

I'll have to learn how to write one but I think I could probably do it :)

Eakam1007 commented:
> I'll have to learn how to write one but I think I could probably do it :)

That's perfect! Do you wanna create an issue for the tests?

sfrunza13 commented:
> > I'll have to learn how to write one but I think I could probably do it :)
> 
> That's perfect! Do you wanna create an issue for the tests?

not super descriptive but #98

humphd reviewed (approved):
Nice work, this looks great.  I'm excited to see the auth happening so soon!
